spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -585,6 +625,54 @@ private void ackImmediate(ConsumerRecord<K, V> record) {
 		}
 
 		private void invokeListener(final ConsumerRecords<K, V> records) {
+			if (this.isBatchListener) {
+				invokeBatchListener(records);
+			}
+			else {
+				invokeRecordListener(records);
+			}
+		}
+
+		private void invokeBatchListener(final ConsumerRecords<K, V> records) {
+			List<ConsumerRecord<K, V>> recordList = new LinkedList<ConsumerRecord<K, V>>();
+			Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
+			while (iterator.hasNext()) {
+				recordList.add(iterator.next());
+			}
+			if (recordList.size() > 0) {
+				try {
+					if (this.batchAcknowledgingMessageListener != null) {
+						this.batchAcknowledgingMessageListener.onMessage(recordList,
+								new ConsumerBatchAcknowledgment(recordList, this.isManualImmediateAck));
+					}
+					else {
+						this.batchListener.onMessage(recordList);
+						if (!this.isAnyManualAck && !this.autoCommit) {
+							for (ConsumerRecord<K, V> record : recordList) {
+								this.acks.put(record);
+							}
+						}
+					}
+				}
+				catch (Exception e) {
+					if (this.containerProperties.isAckOnError() && !this.autoCommit) {
+						for (ConsumerRecord<K, V> record : recordList) {
+							this.acks.add(record);
```

Comment:
I don't remember if I asked similar question to @mbogoevici , when he had been doing the original `ackOnError` solution, but let me ask it again if that.

The main code uses `BlockingQueue.put()` - block/wait instead of this `IllegalState` `add()`.
I guess it doesn't make sense for the used `LinkedBlockingQueue`, but maybe it would be better to be consistent. At least to let me stop to remember such a case and don't ask this question again 😄